spring-kafka/src/main/java/org/springframework/kafka/config/ShareKafkaListenerContainerFactory.java

```diff
@@ -0,0 +1,175 @@
+ * @since 4.0
+ */
+public class ShareKafkaListenerContainerFactory<K, V>
+		implements KafkaListenerContainerFactory<ShareKafkaMessageListenerContainer<K, V>>, ApplicationEventPublisherAware, ApplicationContextAware {
+
+	private final ShareConsumerFactory<? super K, ? super V> shareConsumerFactory;
+
+	private @Nullable Boolean autoStartup;
+
+	private @Nullable Integer phase;
+
+	private @Nullable ApplicationEventPublisher applicationEventPublisher;
+
+	private @Nullable ApplicationContext applicationContext;
+
+	/**
+	 * Construct an instance with the provided consumer factory.
+	 * @param shareConsumerFactory the share consumer factory
+	 */
+	public ShareKafkaListenerContainerFactory(ShareConsumerFactory<K, V> shareConsumerFactory) {
+		this.shareConsumerFactory = shareConsumerFactory;
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) {
+		this.applicationContext = applicationContext;
+	}
+
+	/**
+	 * Set whether containers created by this factory should auto-start.
+	 * @param autoStartup true to auto-start
+	 */
+	public void setAutoStartup(Boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	/**
+	 * Set the phase in which containers created by this factory should start and stop.
+	 * @param phase the phase
+	 */
+	public void setPhase(Integer phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	@Override
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public ShareKafkaMessageListenerContainer<K, V> createListenerContainer(KafkaListenerEndpoint endpoint) {
+		ShareKafkaMessageListenerContainer<K, V> instance = createContainerInstance(endpoint);
+		JavaUtils.INSTANCE
+				.acceptIfNotNull(endpoint.getId(), instance::setBeanName);
+		if (endpoint instanceof AbstractKafkaListenerEndpoint abstractKafkaListenerEndpoint) {
+			configureEndpoint(abstractKafkaListenerEndpoint);
+		}
+		// TODO: No message converter for queue at the moment
+		endpoint.setupListenerContainer(instance, null);
+		initializeContainer(instance, endpoint);
+		return instance;
+	}
+
+	private void configureEndpoint(AbstractKafkaListenerEndpoint<K, V> endpoint) {
+		// Minimal configuration; can add more properties later
+	}
+
+	/**
+	 * Initialize the provided container with common configuration properties.
+	 * @param instance the container instance
+	 * @param endpoint the endpoint
+	 */
+	protected void initializeContainer(ShareKafkaMessageListenerContainer<K, V> instance, KafkaListenerEndpoint endpoint) {
+		ContainerProperties properties = instance.getContainerProperties();
+		JavaUtils.INSTANCE
+				.acceptIfNotNull(endpoint.getAutoStartup(), instance::setAutoStartup)
+				.acceptIfNotNull(this.autoStartup, autoStartup -> {
+					if (endpoint.getAutoStartup() == null) {
+						instance.setAutoStartup(autoStartup);
+					}
+				})
```

Comment:
Some is off with these two calls.
Both of them are about calling `instance.setAutoStartup()` in the end.
Can we calculate a single `autoStartup` variable before propagating it down?